src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs

```diff
@@ -76,5 +84,54 @@ public ModelBindingResult(object model, string key, bool isModelSet, ModelValida
         /// A <see cref="ModelValidationNode"/> associated with the current <see cref="ModelBindingResult"/>.
         /// </summary>
         public ModelValidationNode ValidationNode { get; }
+
+        public override bool Equals(object obj)
+        {
+            var other = obj as ModelBindingResult?;
+            if (other == null)
+            {
+                return false;
+            }
+            else
+            {
+                return Equals(other.Value);
+            }
+        }
+
+        public override int GetHashCode()
+        {
+            if (Key == null)
+            {
+                return 0;
+            }
+            else if (IsModelSet)
+            {
+                Debug.Assert(Key != null);
```

Comment:
Why assert that you're in an `else if` block? Could see asserting `!IsModelSet` in the previous block and `Model == null` here but that's about it.